### PR TITLE
feat: drag n drop and Container jump enhancements

### LINKED ISSUE #12892: [Bug]: Container jump is not smooth when the mouse pointer while moving the widget is away from the container
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

[LOOM DEMO](https://www.loom.com/share/190ce5d5d2584ad4888200a2b49ae398)

The recording on DP but the behaviour is same on production.

### Steps To Reproduce

1. Drag and drop a Container widget
2. Drag and drop a button widget
3. Now click on the button widget near to the container and try to move inside the container
4. Now repeat the same step but keep the mouse pointer away from container

### Public Sample App

_No response_

### Version

Cloud

### LINKED ISSUE #19139: [Task]: Drag and drop Improvements
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

- Resize widgets against Canvas Canvas Boundaries
- Resize Widgets while colliding against Containers/Drop targets

### PR DESCRIPTION
**Pull request by rahulramesha**
## Description

This PR includes Changes for Drag and drop Improvements, That includes,
- Resizing dragging widgets along Container edges.
- Initially resize widgets against Container/Droptarget widgets.
- While holding close to Container/Droptarget widgets for half a second, start to reflow the widget.

Fixes #19139 
Fixes #12892

Media

https://user-images.githubusercontent.com/71900764/209154834-66acecbb-2df8-4598-86d5-4fe7843dd21b.mp4

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- Manual
- Jest

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/src/reflow/reflowUtils.ts [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
 import {
+  BlockSpace,
   CollidingSpace,
   CollidingSpaceMap,
   CollisionAccessors,
   CollisionMap,
   CollisionTree,
   CollisionTreeCache,
   GridProps,
+  HORIZONTAL_RESIZE_LIMIT,
```
Comment (marks0351) [inline app/client/src/reflow/reflowUtils.ts:13]: pls add min to the constants name.

File: app/client/src/reflow/reflowUtils.ts [hunk 2]
```diff
@@ -296,8 +306,33 @@ export function getCollidingSpaceMap(
...
-    for (const occupiedSpace of occupiedSpaces) {
+  let reflowableOccSpaces = [...occupiedSpaces],
+    currSpacePositions = [...newSpacePositions];
+
+  let shouldRegisterContainerTimeout = false;
+
+  //if droptargets are not to be reflowed, resize space positions
+  // and omit drop targets from the spaces
+  if (!shouldReflowDropTarget) {
+    ({
```
Comment (marks0351) [inline app/client/src/reflow/reflowUtils.ts:317]: pls add comments to this.

File: app/client/src/reflow/reflowUtils.ts [hunk 3]
```diff
@@ -1898,3 +1981,290 @@ export function getRelativeCollidingValue(
...
+  const possiblePushDirections = getPossiblePushDirections(
+    containerSpace,
+    mousePosition,
+  );
+
+  if (
+    possiblePushDirections.length < 1 ||
+    possiblePushDirections.includes(currentDirection)
+  )
+    return currentDirection;
```
Comment (marks0351) [inline app/client/src/reflow/reflowUtils.ts:2035]: pls add {}

File: app/client/src/reflow/reflowUtils.ts [hunk 4]
```diff
@@ -1898,3 +1981,290 @@ export function getRelativeCollidingValue(
...
+ * @param direction ReflowDirection
+ * @returns
+ */
+export function modifyResizePosition(
+  newSpacePosition: BlockSpace,
+  collidingContainer: BlockSpace,
+  direction: ReflowDirection,
+): BlockSpace {
+  if (!direction || direction === ReflowDirection.UNSET)
+    return newSpacePosition;
```
Comment (marks0351) [inline app/client/src/reflow/reflowUtils.ts:2192]: {} braces please

File: app/client/src/utils/hooks/useReflow.ts [hunk 5]
```diff
@@ -75,115 +83,210 @@ export const useReflow = (
...
+              prevMovementMap.current = movementMap || {};
+            } else if (isReflowing.current) {
+              isReflowing.current = false;
+              throttledDispatch.cancel();
+              dispatch(stopReflowAction());
+              shouldReflowDropTargets.current = false;
+            }
+          }, 500);
+        } else if (
+          !collidingSpaces.some(
```
Comment (marks0351) [inline app/client/src/utils/hooks/useReflow.ts:247]: check if it is needed.

**Review by marks0351 — CHANGES_REQUESTED** [review]

File: app/client/src/utils/hooks/useReflow.ts (hunk 5 above)
Comment (rahulramesha) [reply app/client/src/utils/hooks/useReflow.ts:248]: This check seems required as it stops from resetting params even when colliding with containers

File: app/client/src/pages/common/CanvasArenas/hooks/canvasDraggingUtils.ts [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    topOffset =
+      topRow + rowHeight > VERTICAL_RESIZE_LIMIT
+        ? topRow
+        : VERTICAL_RESIZE_LIMIT - rowHeight;
+  }
+
+  if (
+    topRow + rowHeight > parentBottomRow &&
+    !canExtend &&
+    fixedHeight === undefined
```
[the remark this answers is not in the document]
Comment (rahulramesha) [reply app/client/src/pages/common/CanvasArenas/hooks/canvasDraggingUtils.ts:247]: yes, if it has dynamic height it will not resize vertically

**Review by marks0351 — DISMISSED** [review]

**Comment by kamakshibhat-appsmith** [comment]
https://www.loom.com/share/9808faae68004df79a9a6d3af5f73f42
The drop target is showing in red even though it is a legit one

**Comment by kamakshibhat-appsmith** [comment]
Looks good @rahulramesha

**Comment by kamakshibhat-appsmith** [comment]
Looks good @rahulramesha

**Comment by kamakshibhat-appsmith** [comment]
few Questions:
1. The size of the input box is slightly increased and there is a scroll introduced inside the input box as soon as data type is changed to multi-line with fixed mode. This also makes the the input text inside the input box look misaligned because the text doesn't seem start from the top (there seems to be extra padding from the top than usual ) [LOOM](https://www.loom.com/share/0f5271f03c1941d98094fb5336c086db)
2. with data-type muti-line text and autoheight enabled, the default rows gets increased for the input type . is this expected? [LOOM](https://www.loom.com/share/6de48668c9124e128c298674c399dd99) 
@ankurrsinghal

**Comment by Wall995501** [comment]
[GE](Wallpaper)

**Pull request merged**